edk2toollib/acpi/ivrs_parser.py

```diff
@@ -0,0 +1,643 @@
+            for ele in val:
+                sum += ele
+        else:
+            if (val & 0xFFFFFFFF00000000) != 0:
+                raise Exception("Invalid 32 bit value: %x", val)
+            else:
+                sum = ((val & 0xFF) + ((val >> 8) & 0xFF) + ((val >> 16) & 0xFF) + ((val >> 24) & 0xFF))
+        return sum
+
+    @staticmethod
+    def byteSum64(val):
+        if (type(val)) is bytes:
+            if len(val) > 8:
+                raise Exception("Invalid 64 bit value: %x", val)
+            sum = 0
+            for ele in val:
+                sum += ele
+        else:
+            sum = (val & 0xFF) + ((val >> 8) & 0xFF) +\
+                  ((val >> 16) & 0xFF) + ((val >> 24) & 0xFF) +\
+                  ((val >> 32) & 0xFF) + ((val >> 40) & 0xFF) +\
+                  ((val >> 48) & 0xFF) + ((val >> 56) & 0xFF)
+        return sum
+
+    @staticmethod
+    def validateChecksum8(data):
+        sum = 0
+        for byte in data:
+            sum += byte
+        return (0x100 - (sum & 0xFF)) & 0xFF
+
+    def addIVHDEntry(self, ivhd):
+        # append raw data, update length and checksum
+        self.data += ivhd.data
+        self.ivrs_table.Length += ivhd.Length
+        self.ivrs_table.SubStructs.append(ivhd)
+        self.ivrs_table.Checksum = self.ivrs_table.calculateACPISum(self.data)
+
+    def addIVMDEntry(self, ivmd):
+        # append raw data, update length and checksum
+        self.data += ivmd.data
+        self.ivrs_table.Length += ivmd.Length
+        self.ivrs_table.IVMDlist.append(ivmd)
+        self.ivrs_table.Checksum = self.ivrs_table.calculateACPISum(self.data)
+
+    def IVRSBitEnabled(self):
+        return bool(self.ivrs_table.IVRSBit)
+
+    def CheckIVMDCount(self, goldenxml=None):
+        goldenignores = list()
+
+        if goldenxml is None or not os.path.isfile(goldenxml):
+            print("XML File not found")
+        else:
+            goldenfile = ET.parse(goldenxml)
+            goldenroot = goldenfile.getroot()
+            for entry in goldenroot:
+                if entry.tag == "IVMD":
+                    goldenignores.append(entry.attrib)
+
+        for IVMD in self.ivrs_table.IVMDlist:
+            if not IVMD.validateIVMD(goldenignores):
+                print("IVMD PCIe Endpoint " + str(IVMD) + " found but not in golden XML")
+                return False
+
+        return True
+
+    class ACPI_TABLE_HEADER(object):
+        struct_format = '=4sIBB6s8sI4sIIQ'
+        size = struct.calcsize(struct_format)
+
+        def __init__(self, header_byte_array):
+            (self.Signature,
+             self.Length,
+             self.Revision,
+             self.Checksum,
+             self.OEMID,
+             self.OEMTableID,
+             self.OEMRevision,
+             self.CreatorID,
+             self.CreatorRevision,
+             self.IVinfo,
+             self.Reserved) = struct.unpack_from(IVRS_TABLE.ACPI_TABLE_HEADER.struct_format, header_byte_array)
```

Comment:
Since none of these specify an offset, let's use unpack() rather than unpack_from().

Reply:
unpack() needs exact format match (including length), some modifications were made to change them to unpack()